File: learning-factory/warehouse/artifacts/projects/build-your-own-x/remediation/27debb2830b4885fa831/repair-v1/job_byox_repair_v1_g1_f7de7abfdff812ca5a1921818f925788/attempt-001/sealed/reference/src/lexer.js
import { LexError, boundedInteger } from "./errors.js";
import { KEYWORDS, TokenType as T } from "./tokens.js";

const MAX_SOURCE_LENGTH = 1_000_000;
const MAX_TOKENS = 200_000;

export function tokenize(source, options = {}) {
  if (typeof source !== "string") throw new TypeError("source must be a string");
  const maxSourceLength = boundedInteger(options, "maxSourceLength", MAX_SOURCE_LENGTH);
  const maxTokens = boundedInteger(options, "maxTokens", MAX_TOKENS);
  if (source.length > maxSourceLength) {
    throw new LexError(`Source exceeds ${maxSourceLength} code units`, { line: 1, column: 1 });
  }
  return new Scanner(source, maxTokens).scan();
}

class Scanner {
  constructor(source, maxTokens) {
    this.source = source;
    this.maxTokens = maxTokens;
    this.tokens = [];
    this.start = 0;
    this.current = 0;
    this.line = 1;
    this.column = 1;
    this.startLine = 1;
    this.startColumn = 1;
  }

  scan() {
    while (!this._atEnd()) {
      this.start = this.current;
      this.startLine = this.line;
      this.startColumn = this.column;
      this._scanToken();
    }
    this.tokens.push({
      type: T.EOF,
      lexeme: "",
      literal: null,
      line: this.line,
      column: this.column
    });
    return this.tokens;
  }

  _scanToken() {
    const c = this._advance();
    const simple = {
      "(": T.LEFT_PAREN,
      ")": T.RIGHT_PAREN,
      "{": T.LEFT_BRACE,
      "}": T.RIGHT_BRACE,
      ";": T.SEMICOLON,
      "+": T.PLUS,
      "-": T.MINUS,
      "*": T.STAR
    };
    if (simple[c]) return this._add(simple[c]);

    switch (c) {
      case "!": return this._add(this._match("=") ? T.BANG_EQUAL : T.BANG);
      case "=": return this._add(this._match("=") ? T.EQUAL_EQUAL : T.EQUAL);
      case "<": return this._add(this._match("=") ? T.LESS_EQUAL : T.LESS);
      case ">": return this._add(this._match("=") ? T.GREATER_EQUAL : T.GREATER);
      case "/":
        if (this._match("/")) {
          while (!this._atEnd() && this._peek() !== "\n" && this._peek() !== "\r") this._advance();
          return;
        }
        return this._add(T.SLASH);
      case " ":
      case "\t":
        return;
      case "\n":
        return;
      case "\r":
        if (this._peek() === "\n") {
          this._advance();
        } else {
          this.line += 1;
          this.column = 1;
        }
        return;
      case '"': return this._string();
      default:
        if (isDigit(c)) return this._number();
        if (isAlpha(c)) return this._identifier();
        throw new LexError(`Unexpected character ${JSON.stringify(c)}`, this._startLocation());
    }
  }

  _string() {
    let value = "";
    while (!this._atEnd()) {
      const location = { line: this.line, column: this.column };
      const c = this._advance();
      if (c === '"') {
        this._add(T.STRING, value);
        return;
      }
      if (c === "\n" || c === "\r") {
        throw new LexError("Raw line break in string", location);
      }
      if (c !== "\\") {
        value += c;
        continue;
      }
      if (this._atEnd()) throw new LexError("Unterminated string", this._startLocation());
      const escapeLocation = { line: this.line, column: this.column };
      const escaped = this._advance();
      const escapes = { '"': '"', "\\": "\\", n: "\n", r: "\r", t: "\t" };
      if (!Object.prototype.hasOwnProperty.call(escapes, escaped)) {
        throw new LexError(`Unsupported escape \\${escaped}`, escapeLocation);
      }
      value += escapes[escaped];
    }
    throw new LexError("Unterminated string", this._startLocation());
  }

  _number() {
    while (isDigit(this._peek())) this._advance();
    if (this._peek() === "." && isDigit(this._peekNext())) {
      this._advance();
      while (isDigit(this._peek())) this._advance();
    }
    const value = Number(this.source.slice(this.start, this.current));
    if (!Number.isFinite(value)) throw new LexError("Number is not finite", this._startLocation());
    this._add(T.NUMBER, value);
  }

  _identifier() {
    while (isAlphaNumeric(this._peek())) this._advance();
    const text = this.source.slice(this.start, this.current);
    const type = Object.prototype.hasOwnProperty.call(KEYWORDS, text)
      ? KEYWORDS[text]
      : T.IDENTIFIER;
    this._add(type);
  }

  _add(type, literal = null) {
    if (this.tokens.length >= this.maxTokens) {
      throw new LexError(`Token count exceeds ${this.maxTokens}`, this._startLocation());
    }
    this.tokens.push({
      type,
      lexeme: this.source.slice(this.start, this.current),
      literal,
      line: this.startLine,
      column: this.startColumn
    });
  }

  _advance() {
    const c = this.source[this.current++];
    if (c === "\n") {
      this.line += 1;
      this.column = 1;
    } else {
      this.column += 1;
    }
    return c;
  }

  _match(expected) {
    if (this._atEnd() || this.source[this.current] !== expected) return false;
    this._advance();
    return true;
  }

  _peek() { return this._atEnd() ? "\0" : this.source[this.current]; }
  _peekNext() { return this.current + 1 >= this.source.length ? "\0" : this.source[this.current + 1]; }
  _atEnd() { return this.current >= this.source.length; }
  _startLocation() { return { line: this.startLine, column: this.startColumn }; }
}

function isDigit(c) { return c >= "0" && c <= "9"; }
function isAlpha(c) { return (c >= "a" && c <= "z") || (c >= "A" && c <= "Z") || c === "_"; }
function isAlphaNumeric(c) { return isAlpha(c) || isDigit(c); }

File: learning-factory/warehouse/artifacts/projects/build-your-own-x/programming-language/crafting-interpreters-a-handbook-for-making-programming-languages-40bc68c012/job_byox_build_s2_d5738a92867d38332150840e7d7f6214/attempt-001/sealed/reference/src/main/java/org/learningfactory/mica/Lexer.java
package org.learningfactory.mica;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public final class Lexer {
    private static final Map<String, TokenType> KEYWORDS = Map.ofEntries(
            Map.entry("and", TokenType.AND),
            Map.entry("else", TokenType.ELSE),
            Map.entry("false", TokenType.FALSE),
            Map.entry("if", TokenType.IF),
            Map.entry("let", TokenType.LET),
            Map.entry("nil", TokenType.NIL),
            Map.entry("or", TokenType.OR),
            Map.entry("print", TokenType.PRINT),
            Map.entry("true", TokenType.TRUE),
            Map.entry("while", TokenType.WHILE));

    private final String source;
    private final List<Token> tokens = new ArrayList<>();
    private int start;
    private int current;
    private int line = 1;
    private int column = 1;
    private int startLine;
    private int startColumn;
    private List<Token> scanned;

    public Lexer(String source) {
        if (source == null) {
            throw new MicaException(MicaException.Kind.LEX, 1, 1, "source must not be null");
        }
        this.source = source;
    }

    public List<Token> scanTokens() {
        if (scanned != null) return scanned;
        while (!isAtEnd()) {
            start = current;
            startLine = line;
            startColumn = column;
            scanToken();
        }
        tokens.add(new Token(TokenType.EOF, "", null, line, column));
        scanned = List.copyOf(tokens);
        return scanned;
    }

    private void scanToken() {
        char c = advance();
        switch (c) {
            case '(' -> add(TokenType.LEFT_PAREN);
            case ')' -> add(TokenType.RIGHT_PAREN);
            case '{' -> add(TokenType.LEFT_BRACE);
            case '}' -> add(TokenType.RIGHT_BRACE);
            case ',' -> add(TokenType.COMMA);
            case '-' -> add(TokenType.MINUS);
            case '+' -> add(TokenType.PLUS);
            case ';' -> add(TokenType.SEMICOLON);
            case '*' -> add(TokenType.STAR);
            case '!' -> add(match('=') ? TokenType.BANG_EQUAL : TokenType.BANG);
            case '=' -> add(match('=') ? TokenType.EQUAL_EQUAL : TokenType.EQUAL);
            case '<' -> add(match('=') ? TokenType.LESS_EQUAL : TokenType.LESS);
            case '>' -> add(match('=') ? TokenType.GREATER_EQUAL : TokenType.GREATER);
            case '/' -> {
                if (match('/')) {
                    while (peek() != '\n' && !isAtEnd()) advance();
                } else {
                    add(TokenType.SLASH);
                }
            }
            case ' ', '\r', '\t', '\n' -> { }
            case '"' -> string();
            default -> {
                if (isDigit(c)) number();
                else if (isIdentifierStart(c)) identifier();
                else throw lex(startLine, startColumn, "unexpected character '" + printable(c) + "'");
            }
        }
    }

    private void string() {
        StringBuilder value = new StringBuilder();
        while (!isAtEnd()) {
            char c = advance();
            if (c == '"') {
                add(TokenType.STRING, value.toString());
                return;
            }
            if (c != '\\') {
                value.append(c);
                continue;
            }

            int escapeLine = line;
            int escapeColumn = column - 1;
            if (isAtEnd()) {
                throw lex(startLine, startColumn, "unterminated string");
            }
            char escaped = advance();
            switch (escaped) {
                case 'n' -> value.append('\n');
                case 't' -> value.append('\t');
                case '"' -> value.append('"');
                case '\\' -> value.append('\\');
                default -> throw lex(escapeLine, escapeColumn,
                        "invalid escape '\\" + printable(escaped) + "'");
            }
        }
        throw lex(startLine, startColumn, "unterminated string");
    }

    private void number() {
        while (isDigit(peek())) advance();
        if (peek() == '.' && isDigit(peekNext())) {
            advance();
            while (isDigit(peek())) advance();
        }

        String text = source.substring(start, current);
        try {
            double value = Double.parseDouble(text);
            if (!Double.isFinite(value)) {
                throw lex(startLine, startColumn, "number literal is not finite");
            }
            add(TokenType.NUMBER, value);
        } catch (NumberFormatException exception) {
            throw lex(startLine, startColumn, "invalid number literal");
        }
    }

    private void identifier() {
        while (isIdentifierPart(peek())) advance();
        String text = source.substring(start, current);
        add(KEYWORDS.getOrDefault(text, TokenType.IDENTIFIER));
    }

    private boolean match(char expected) {
        if (isAtEnd() || source.charAt(current) != expected) return false;
        advance();
        return true;
    }

    private char peek() {
        return isAtEnd() ? '\0' : source.charAt(current);
    }

    private char peekNext() {
        return current + 1 >= source.length() ? '\0' : source.charAt(current + 1);
    }

    private char advance() {
        char c = source.charAt(current++);
        if (c == '\n') {
            line++;
            column = 1;
        } else {
            column++;
        }
        return c;
    }

    private boolean isAtEnd() {
        return current >= source.length();
    }

    private void add(TokenType type) {
        add(type, null);
    }

    private void add(TokenType type, Object literal) {
        tokens.add(new Token(type, source.substring(start, current), literal, startLine, startColumn));
    }

    private static boolean isDigit(char c) {
        return c >= '0' && c <= '9';
    }

    private static boolean isIdentifierStart(char c) {
        return c == '_' || (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z');
    }

    private static boolean isIdentifierPart(char c) {
        return isIdentifierStart(c) || isDigit(c);
    }

    private static String printable(char c) {
        return switch (c) {
            case '\n' -> "\\n";
            case '\r' -> "\\r";
            case '\t' -> "\\t";
            default -> Character.toString(c);
        };
    }

    private static MicaException lex(int line, int column, String detail) {
        return new MicaException(MicaException.Kind.LEX, line, column, detail);
    }
}
